### PR TITLE
Sage For Node Classification

### PR DESCRIPTION
**Pull request by jalencato**
*Issue #, if available:*

*Description of changes:*

Add sage for node classification

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/data/ogbn_datasets.py [hunk 1]
```diff
@@ -52,13 +52,16 @@ class OGBTextFeatDataset(GSgnnTextDataset):
         Whether to use the original features or the bert generated ones
     bert_model_name : String
         The bert model used for tokenization
+    is_homo: Boolean
+        If we want to generate a homogeneous graph to compat DGL
```
Comment (classicsong) [inline python/graphstorm/data/ogbn_datasets.py:56]:
```suggestion
        If we want to generate a homogeneous graph.
```

File: python/graphstorm/data/ogbn_datasets.py [hunk 2]
```diff
@@ -176,58 +187,58 @@ def process(self):
...
                                                                       dtype=th.bool)
-                g.edges["rev-interacts"].data['val_mask'] = th.full((int_edges,), False,
+                g.edges[self.rev_edge_type].data['val_mask'] = th.full((int_edges,), False,
                                                                     dtype=th.bool)
-                g.edges["rev-interacts"].data['test_mask'] = th.full((int_edges,), False,
+                g.edges[self.rev_edge_type].data['test_mask'] = th.full((int_edges,), False,
                                                                      dtype=th.bool)
-                g.edges["rev-interacts"].data['train_mask'][: int(int_edges * train_pct)] = True
-                g.edges["rev-interacts"].data['val_mask'][int(int_edges*train_pct):
+                g.edges[self.rev_edge_type].data['train_mask'][: int(int_edges * train_pct)] = True
```
Comment (classicsong) [inline python/graphstorm/data/ogbn_datasets.py:225]: This is not correct when the graph is homogeneous. When it is a homogeneous graph,`g.edges[self.edge_type]` and `g.edges[self.rev_edge_type]` are identical . So you are setting the same set of edges to True.
Can you check whether `(self.node_type, self.rev_edge_type, self.node_type): (dst, src)` is appended at the end of `(self.node_type, self.edge_type, self.node_type): (src, dst)`.

File: python/graphstorm/gsf.py [hunk 3]
```diff
@@ -32,6 +32,7 @@
 from .model.lm_embed import GSLMNodeEncoderInputLayer, GSPureLMNodeInputLayer
 from .model.rgcn_encoder import RelationalGCNEncoder
 from .model.rgat_encoder import RelationalGATEncoder
+from .model.sage_encoder import SAGEEncoder
```
Comment (classicsong) [inline python/graphstorm/gsf.py:35]: You also need to update the init.py

File: python/graphstorm/gsf.py [hunk 4]
```diff
@@ -476,10 +477,29 @@ def set_encoder(model, g, config, train_task):
...
+                                num_ffn_layers_in_gnn=config.num_ffn_layers_in_gnn)
     else:
         assert False, "Unknown gnn model type {}".format(model_encoder_type)
     model.set_gnn_encoder(gnn_encoder)
 
+
+def check_homo(g):
+    if g.ntypes == ['_N'] and g.etypes == ['_E']:
+        return True
+    print("It is not a required homogeneous graph for DGL, the node type should be '_N' "
```
Comment (classicsong) [inline python/graphstorm/gsf.py:498]: Remove this print

File: python/graphstorm/model/sage_encoder.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        self.aggregator_type = aggregator_type
+
+        self.conv = dglnn.SAGEConv(self.in_feat, self.out_feat, self.aggregator_type,
+                                   feat_drop=dropout, bias=bias)
+
+        self.activation = activation
+        # ngnn
+        self.num_ffn_layers_in_gnn = num_ffn_layers_in_gnn
+        self.ngnn_mlp = NGNNMLP(out_feat, out_feat,
+                                     num_ffn_layers_in_gnn, ffn_activation, dropout)
```
Comment (classicsong) [inline python/graphstorm/model/sage_encoder.py:68]:
```suggestion
                                 num_ffn_layers_in_gnn, ffn_activation, dropout)
```

File: python/graphstorm/model/sage_encoder.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+    def forward(self, g, inputs):
+        """Forward computation
+
+        Parameters
+        ----------
+        g : DGLHeteroGraph
+            Input graph.
+        inputs : dict["_N", torch.Tensor]
```
Comment (classicsong) [inline python/graphstorm/model/sage_encoder.py:77]: dict{"_N", torch.Tensor} ?

File: python/graphstorm/model/sage_encoder.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    """
+    def __init__(self,
+                 h_dim, out_dim,
+                 num_hidden_layers=1,
+                 dropout=0,
+                 aggregator_type='mean',
+                 activation=F.relu,
+                 num_ffn_layers_in_gnn=0):
+        super(SAGEEncoder, self).__init__(h_dim, out_dim, num_hidden_layers)
+        self.aggregator_type = aggregator_type
```
Comment (classicsong) [inline python/graphstorm/model/sage_encoder.py:120]: Actually, we do not need to save aggregator_type and dropout as the fields of SAGEEncoder.

File: python/graphstorm/model/sage_encoder.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        self.dropout = dropout
+
+        self.layers = nn.ModuleList()
+        for _ in range(num_hidden_layers):
+            self.layers.append(SAGEConv(h_dim, h_dim, self.aggregator_type,
+                                        bias=False, activation=activation,
+                                        num_ffn_layers_in_gnn=num_ffn_layers_in_gnn))
+
```
Comment (classicsong) [inline python/graphstorm/model/sage_encoder.py:128]: We also need to pass dropout into SAGEConv

File: python/graphstorm/model/sage_encoder.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        self.layers.append(SAGEConv(h_dim, out_dim, self.aggregator_type,
+                                    bias=False, activation=activation,
+                                    num_ffn_layers_in_gnn=num_ffn_layers_in_gnn))
+
+    # TODO(zhengda) refactor this to support edge features.
```
Comment (classicsong) [inline python/graphstorm/model/sage_encoder.py:133]:
```suggestion
    # TODO refactor this to support edge features.
```

File: tests/unit-tests/test_gnn.py [hunk 10]
```diff
@@ -134,7 +153,7 @@ def require_cache_embed(self):
         assert_almost_equal(embs3[ntype][0:len(embs3[ntype])].numpy(),
                             embs4[ntype][0:len(embs4[ntype])].numpy())
 
-    target_nidx = {"n1": th.arange(g.number_of_nodes("n0"))}
+    target_nidx = {"n1": th.arange(g.number_of_nodes("n0"))} if not is_homo else {"_N": th.arange(g.number_of_nodes("_N"))}
```
Comment (classicsong) [inline tests/unit-tests/test_gnn.py:156]: Can you make each line shorter than 100 characters?

File: tests/unit-tests/test_gnn.py [hunk 11]
```diff
@@ -848,4 +891,4 @@ def test_link_prediction_weight():
 
     test_mlp_edge_prediction()
     test_mlp_node_prediction()
-    test_mlp_link_prediction()
+    test_mlp_link_prediction()
```
Comment (classicsong) [inline tests/unit-tests/test_gnn.py:894]: keep the newline at the end of the file.

**Review by classicsong — COMMENTED** [review]
Can you add one test in test.sh for Sage model as an end2end test?

File: python/graphstorm/data/ogbn_datasets.py (hunk 2 above)
Comment (jalencato) [reply python/graphstorm/data/ogbn_datasets.py:225]: The logic here should be right. I have already added the `(self.node_type, self.rev_edge_type, self.node_type): (dst, src)` when adding reverse edge option. So even if they are set the same set of edges to True, it should be fine, we expect to do so. I add a if condition here to not generate the mask again when it is homogeneous graph. I think then it would be fine.

File: python/graphstorm/gsf.py [hunk 12]
```diff
@@ -481,10 +482,27 @@ def set_encoder(model, g, config, train_task):
                                            dropout=dropout,
                                            use_self_loop=config.use_self_loop,
                                            num_ffn_layers_in_gnn=config.num_ffn_layers_in_gnn)
+    elif model_encoder_type == "sage":
+        # we need to check if the graph is homogeneous
+        assert check_homo(g) == True
```
Comment (classicsong) [inline python/graphstorm/gsf.py:487]: Please add an error message as assert xxx, "Why assert fails."

File: python/graphstorm/gsf.py [hunk 13]
```diff
@@ -481,10 +482,27 @@ def set_encoder(model, g, config, train_task):
...
+        # we need to set the num_layers -1 because there is an output layer that is hard coded.
+        gnn_encoder = SAGEEncoder(h_dim=config.hidden_size,
+                                out_dim=config.hidden_size,
```
Comment (classicsong) [inline python/graphstorm/gsf.py:490]: indent

File: python/graphstorm/gsf.py [hunk 14]
```diff
@@ -481,10 +482,27 @@ def set_encoder(model, g, config, train_task):
...
+                                num_hidden_layers=config.num_layers - 1,
+                                dropout=dropout,
+                                aggregator_type='pool',
+                                num_ffn_layers_in_gnn=config.num_ffn_layers_in_gnn)
     else:
         assert False, "Unknown gnn model type {}".format(model_encoder_type)
     model.set_gnn_encoder(gnn_encoder)
 
+
+def check_homo(g):
```
Comment (classicsong) [inline python/graphstorm/gsf.py:500]: Please add docstring to this function.

File: python/graphstorm/model/sage_encoder.py [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+                 num_ffn_layers_in_gnn=0,
+                 ffn_activation=F.relu):
+        super(SAGEConv, self).__init__()
+        self.in_feat, self.out_feat = in_feat, out_feat
+        self.aggregator_type = aggregator_type
+
+        self.conv = dglnn.SAGEConv(self.in_feat, self.out_feat, self.aggregator_type,
+                                   feat_drop=dropout, bias=bias)
+
+        self.activation = activation
```
Comment (classicsong) [inline python/graphstorm/model/sage_encoder.py:64]: Shall we add a default activation function here?

File: tools/README.md [hunk 16]
```diff
@@ -67,4 +67,18 @@ python3 /graphstorm/tools/partition_graph_lp.py --dataset ogbn-mag \
...
-For details of the arguments of the two scripts, please refer to their **main()** funciton.
\ No newline at end of file
+For details of the arguments of the two scripts, please refer to their **main()** funciton.
+
+Below command can generate the homogeneous version of ogbn-arxiv. The partitioned graph is save at the /tmp/ogbn_arxiv_nc_2p folder.
+```bash
+python3 /graphstorm/tools/gen_ogb_dataset.py --savepath /tmp/ogbn-arxiv-nc/  \
+                          --dataset ogbn-arxiv \
+                          --retain-original-features true \
+                          --is-homo
```
Comment (classicsong) [inline tools/README.md:77]: Add selfloop

**Review by classicsong — COMMENTED** [review]
Can we add a end2end test in end2end-tests?

**Review by classicsong — APPROVED** [review]
